// @solana/mpp/dist/server/Charge.js
import { address, isTransactionPartialSigner } from '@solana/kit';
import { findAssociatedTokenPda } from '@solana-program/token';
import { Method, Receipt, Store } from 'mppx';
import { DEFAULT_RPC_URLS, TOKEN_2022_PROGRAM, TOKEN_PROGRAM } from '../constants.js';
import * as Methods from '../Methods.js';
import { coSignBase64Transaction } from '../utils/transactions.js';
/**
 * Creates a Solana `charge` method for usage on the server.
 *
 * Supports two settlement modes:
 *
 * - **Pull mode** (`type="transaction"`, default): The server receives a
 *   signed transaction from the client, broadcasts it to Solana, confirms
 *   it, and verifies the transfer on-chain. When `signer` is provided,
 *   the server co-signs as fee payer before broadcasting.
 *
 * - **Push mode** (`type="signature"`): The client has already broadcast
 *   and confirmed the transaction. The server verifies the transfer
 *   on-chain using the signature.
 *
 * @example
 * ```ts
 * import { Mppx, solana } from 'solana-mpp-sdk/server'
 *
 * const mppx = Mppx.create({
 *   methods: [solana.charge({
 *     recipient: 'RecipientPubkey...',
 *     spl: 'EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v',
 *     decimals: 6,
 *     network: 'devnet',
 *   })],
 * })
 *
 * export async function handler(request: Request) {
 *   const result = await mppx.charge({ amount: '1000000', currency: 'USDC' })(request)
 *   if (result.status === 402) return result.challenge
 *   return result.withReceipt(Response.json({ data: '...' }))
 * }
 * ```
 */
export function charge(parameters) {
    const { recipient, currency, decimals, tokenProgram = TOKEN_PROGRAM, network = 'mainnet-beta', store = Store.memory(), splits, signer, } = parameters;
    const isSplToken = currency !== undefined && currency !== 'sol';
    const rpcUrl = parameters.rpcUrl ?? DEFAULT_RPC_URLS[network] ?? DEFAULT_RPC_URLS['mainnet-beta'];
    if (isSplToken && decimals === undefined) {
        throw new Error('decimals is required when currency is a token mint address');
    }
    if (splits && splits.length > 8) {
        throw new Error('splits cannot exceed 8 entries');
    }
    if (signer && !isTransactionPartialSigner(signer)) {
        throw new Error('signer must implement signTransactions() for fee payer mode (e.g. KeyPairSigner, SolanaSigner)');
    }
    return Method.toServer(Methods.charge, {
        defaults: {
            currency: currency ?? 'sol',
            methodDetails: {},
            recipient: '',
        },
        async request({ credential, request }) {
            if (credential) {
                return credential.challenge.request;
            }
            // Pre-fetch a recent blockhash so the client can skip an RPC call.
            let recentBlockhash;
            try {
                const res = await fetch(rpcUrl, {
                    body: JSON.stringify({
                        id: 1,
                        jsonrpc: '2.0',
                        method: 'getLatestBlockhash',
                        params: [{ commitment: 'confirmed' }],
                    }),
                    headers: { 'Content-Type': 'application/json' },
                    method: 'POST',
                });
                const data = (await res.json());
                recentBlockhash = data.result?.value?.blockhash;
            }
            catch {
                // Non-fatal — client will fetch its own blockhash.
            }
            return {
                ...request,
                methodDetails: {
                    network,
                    ...(isSplToken ? { decimals, tokenProgram } : {}),
                    ...(signer ? { feePayer: true, feePayerKey: signer.address } : {}),
                    ...(splits?.length ? { splits } : {}),
                    ...(recentBlockhash ? { recentBlockhash } : {}),
                },
                recipient,
            };
        },
        async verify({ credential }) {
            const cred = credential;
            const challenge = cred.challenge.request;
            const payloadType = resolvePayloadType(cred.payload);
            // Spec: type="signature" MUST NOT be used with feePayer: true
            if (payloadType === 'signature' && challenge.methodDetails.feePayer) {
                throw new Error('type="signature" credentials cannot be used with fee sponsorship (feePayer: true)');
            }
            if (payloadType === 'transaction') {
                return await verifyTransaction(cred, challenge, rpcUrl, recipient, store, signer);
            }
            return await verifySignature(cred, challenge, rpcUrl, recipient, store);
        },
    });
}
// ── Payload type resolution ──
function resolvePayloadType(payload) {
    if (payload.type === 'signature')
        return 'signature';
    if (payload.type === 'transaction')
        return 'transaction';
    throw new Error('Missing or invalid payload type: must be "transaction" or "signature"');
}
// ── Pull mode (type="transaction") ──
async function verifyTransaction(credential, challenge, rpcUrl, recipient, store, signer) {
    const { transaction: clientTxBase64 } = credential.payload;
    if (!clientTxBase64) {
        throw new Error('Missing transaction data in credential payload');
    }
    let txToSend = clientTxBase64;
    // When the server has a fee payer signer, decode the client's partially-signed
    // transaction, add the fee payer signature, and re-encode.
    // NOTE: The fee payer covers all tx costs including ATA rent (~0.002 SOL).
    // Recipients can close ATAs to reclaim rent, forcing re-creation on the next
    // payment. Servers should verify ATA existence before signing or factor rent
    // into pricing to mitigate this drain vector.
    if (signer) {
        txToSend = await coSignBase64Transaction(signer, clientTxBase64);
    }
    // Simulate before broadcast to catch failures without wasting fees.
    await simulateTransaction(rpcUrl, txToSend);
    // Broadcast the (now fully-signed) transaction.
    const signature = await broadcastTransaction(rpcUrl, txToSend);
    // Wait for on-chain confirmation.
    await waitForConfirmation(rpcUrl, signature);
    // Verify the confirmed transaction matches the challenge.
    await verifyOnChain(rpcUrl, signature, challenge, recipient);
    // Mark consumed to prevent replay.
    await store.put(`solana-charge:consumed:${signature}`, true);
    return Receipt.from({
        method: 'solana',
        reference: signature,
        status: 'success',
        timestamp: new Date().toISOString(),
    });
}
// ── Push mode (type="signature") ──
async function verifySignature(credential, challenge, rpcUrl, recipient, store) {
    const { signature } = credential.payload;
    if (!signature) {
        throw new Error('Missing signature in credential payload');
    }
    // Replay prevention: reject already-consumed transaction signatures.
    const consumedKey = `solana-charge:consumed:${signature}`;
    if (await store.get(consumedKey)) {
        throw new Error('Transaction signature already consumed');
    }
    // Fetch and verify the transaction on-chain.
    const tx = await fetchTransaction(rpcUrl, signature);
    if (!tx)
        throw new Error('Transaction not found or not yet confirmed');
    if (tx.meta?.err)
        throw new Error('Transaction failed on-chain');
    const instructions = tx.transaction.message.instructions;
    await verifyInstructions(instructions, challenge, recipient);
    // Mark consumed to prevent replay.
    await store.put(consumedKey, true);
    return Receipt.from({
        method: 'solana',
        reference: signature,
        status: 'success',
        timestamp: new Date().toISOString(),
    });
}
// ── Shared on-chain verification ──
async function verifyOnChain(rpcUrl, signature, challenge, recipient) {
    const tx = await fetchTransaction(rpcUrl, signature);
    if (!tx)
        throw new Error('Transaction not found or not yet confirmed');
    if (tx.meta?.err)
        throw new Error('Transaction failed on-chain');
    const instructions = tx.transaction.message.instructions;
    await verifyInstructions(instructions, challenge, recipient);
}
async function verifyInstructions(instructions, challenge, recipient) {
    const splits = challenge.methodDetails.splits ?? [];
    const splitsTotal = splits.reduce((sum, s) => sum + BigInt(s.amount), 0n);
    const primaryAmount = BigInt(challenge.amount) - splitsTotal;
    if (primaryAmount <= 0n) {
        throw new Error('Splits consume the entire amount — primary recipient must receive a positive amount');
    }
    const mint = challenge.currency !== 'sol' ? challenge.currency : undefined;
    if (mint) {
        // ── SPL token transfers verification ──
        const transfers = instructions.filter(ix => ix.parsed?.type === 'transferChecked' &&
            (ix.programId === TOKEN_PROGRAM || ix.programId === TOKEN_2022_PROGRAM));
        const expectedTokenProgram = challenge.methodDetails.tokenProgram || TOKEN_PROGRAM;
        // Verify primary transfer to recipient.
        await verifySplTransfer(transfers, recipient, String(primaryAmount), mint, expectedTokenProgram);
        // Verify each split transfer.
        for (const split of splits) {
            await verifySplTransfer(transfers, split.recipient, split.amount, mint, expectedTokenProgram);
        }
    }
    else {
        // ── Native SOL transfers verification ──
        const transfers = instructions.filter(ix => ix.parsed?.type === 'transfer' && ix.program === 'system');
        // Verify primary transfer to recipient.
        verifySolTransfer(transfers, recipient, String(primaryAmount));
        // Verify each split transfer.
        for (const split of splits) {
            verifySolTransfer(transfers, split.recipient, split.amount);
        }
    }
}
async function verifySplTransfer(transfers, recipientAddress, expectedAmount, spl, tokenProgram) {
    const [expectedAta] = await findAssociatedTokenPda({
        mint: address(spl),
        owner: address(recipientAddress),
        tokenProgram: address(tokenProgram),
    });
    const index = transfers.findIndex(ix => {
        const info = ix.parsed.info;
        return info.destination === expectedAta && info.mint === spl;
    });
    if (index === -1) {
        throw new Error(`No TransferChecked instruction found for recipient ${recipientAddress}`);
    }
    const [transfer] = transfers.splice(index, 1);
    const info = transfer.parsed.info;
    if (info.tokenAmount.amount !== expectedAmount) {
        throw new Error(`Amount mismatch for ${recipientAddress}: expected ${expectedAmount}, got ${info.tokenAmount.amount}`);
    }
}
function verifySolTransfer(transfers, recipientAddress, expectedAmount) {
    const index = transfers.findIndex(ix => {
        const info = ix.parsed.info;
        return info.destination === recipientAddress;
    });
    if (index === -1) {
        throw new Error(`No system transfer instruction found for recipient ${recipientAddress}`);
    }
    const [transfer] = transfers.splice(index, 1);
    const info = transfer.parsed.info;
    if (String(info.lamports) !== expectedAmount) {
        throw new Error(`Amount mismatch for ${recipientAddress}: expected ${expectedAmount} lamports, got ${info.lamports}`);
    }
}
// ── RPC helpers ──
async function fetchTransaction(rpcUrl, signature) {
    const response = await fetch(rpcUrl, {
        body: JSON.stringify({
            id: 1,
            jsonrpc: '2.0',
            method: 'getTransaction',
            params: [
                signature,
                {
                    commitment: 'confirmed',
                    encoding: 'jsonParsed',
                    maxSupportedTransactionVersion: 0,
                },
            ],
        }),
        headers: { 'Content-Type': 'application/json' },
        method: 'POST',
    });
    const data = (await response.json());
    if (data.error)
        throw new Error(`RPC error: ${data.error.message}`);
    return data.result ?? null;
}
async function simulateTransaction(rpcUrl, base64Tx) {
    const response = await fetch(rpcUrl, {
        body: JSON.stringify({
            id: 1,
            jsonrpc: '2.0',
            method: 'simulateTransaction',
            params: [base64Tx, { commitment: 'confirmed', encoding: 'base64' }],
        }),
        headers: { 'Content-Type': 'application/json' },
        method: 'POST',
    });
    const data = (await response.json());
    if (data.error)
        throw new Error(`RPC error: ${data.error.message}`);
    const simErr = data.result?.value?.err;
    if (simErr) {
        const logs = data.result?.value?.logs ?? [];
        console.error('[solana-mpp] Simulation failed:', JSON.stringify(simErr));
        for (const log of logs)
            console.error('[solana-mpp]', log);
        throw new Error(`Transaction simulation failed: ${JSON.stringify(simErr)}`);
    }
}
async function broadcastTransaction(rpcUrl, base64Tx) {
    const response = await fetch(rpcUrl, {
        body: JSON.stringify({
            id: 1,
            jsonrpc: '2.0',
            method: 'sendTransaction',
            params: [base64Tx, { encoding: 'base64', skipPreflight: false }],
        }),
        headers: { 'Content-Type': 'application/json' },
        method: 'POST',
    });
    const data = (await response.json());
    if (data.error)
        throw new Error(`RPC error: ${data.error.message}`);
    if (!data.result)
        throw new Error('No signature returned from sendTransaction');
    return data.result;
}
async function waitForConfirmation(rpcUrl, signature, timeoutMs = 30_000) {
    const start = Date.now();
    while (Date.now() - start < timeoutMs) {
        const response = await fetch(rpcUrl, {
            body: JSON.stringify({
                id: 1,
                jsonrpc: '2.0',
                method: 'getSignatureStatuses',
                params: [[signature]],
            }),
            headers: { 'Content-Type': 'application/json' },
            method: 'POST',
        });
        const data = (await response.json());
        const status = data.result?.value?.[0];
        if (status) {
            if (status.err) {
                throw new Error(`Transaction failed: ${JSON.stringify(status.err)}`);
            }
            if (status.confirmationStatus === 'confirmed' || status.confirmationStatus === 'finalized') {
                return;
            }
        }
        await new Promise(r => setTimeout(r, 2_000));
    }
    throw new Error('Transaction confirmation timeout');
}
