// ox/_esm/tempo/VirtualMaster.js
import * as Address from '../core/Address.js';
import * as Bytes from '../core/Bytes.js';
import * as Errors from '../core/Errors.js';
import * as Hash from '../core/Hash.js';
import * as Hex from '../core/Hex.js';
import * as TempoAddress from './TempoAddress.js';
import * as VirtualAddress from './VirtualAddress.js';
const tip20Prefix = '0x20c000000000000000000000';
const zeroAddress = '0x0000000000000000000000000000000000000000';
/**
 * Computes the TIP-1022 registration hash for a master address and salt.
 *
 * [TIP-1022](https://docs.tempo.xyz/protocol/tips/tip-1022)
 *
 * The registration hash is `keccak256(masterAddress || salt)` where `salt`
 * is encoded as a 32-byte value.
 *
 * Master addresses must satisfy TIP-1022 registration constraints: they cannot
 * be the zero address, another virtual address, or a TIP-20 token address.
 *
 * @example
 * ```ts twoslash
 * import { VirtualMaster } from 'ox/tempo'
 *
 * const hash = VirtualMaster.getRegistrationHash({
 *   address: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266',
 *   salt: '0x00000000000000000000000000000000000000000000000000000000abf52baf',
 * })
 *
 * hash
 * // @log: '0x0000000058e21090d8f4bee424b90cddc2378aefa1bbbfa1443631a929ae966d'
 * ```
 *
 * @param value - Master address and salt.
 * @returns The registration hash.
 */
export function getRegistrationHash(value) {
    return Hash.keccak256(Hex.concat(resolveAddress(value.address), toFixedHex(value.salt, 32)));
}
/**
 * Derives the 4-byte TIP-1022 `masterId` from a master address and salt.
 *
 * [TIP-1022](https://docs.tempo.xyz/protocol/tips/tip-1022)
 *
 * This returns bytes `[4:8]` of the registration hash, regardless of whether the
 * salt satisfies the proof-of-work requirement.
 *
 * Master addresses must satisfy TIP-1022 registration constraints: they cannot
 * be the zero address, another virtual address, or a TIP-20 token address.
 *
 * @example
 * ```ts twoslash
 * import { VirtualMaster } from 'ox/tempo'
 *
 * const masterId = VirtualMaster.getMasterId({
 *   address: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266',
 *   salt: '0x00000000000000000000000000000000000000000000000000000000abf52baf',
 * })
 *
 * masterId
 * // @log: '0x58e21090'
 * ```
 *
 * @param value - Master address and salt.
 * @returns The derived master identifier.
 */
export function getMasterId(value) {
    return Hex.slice(getRegistrationHash(value), 4, 8);
}
/**
 * Validates that a salt satisfies the TIP-1022 32-bit proof-of-work requirement.
 *
 * [TIP-1022](https://docs.tempo.xyz/protocol/tips/tip-1022)
 *
 * Returns `false` for invalid master addresses, including the zero address,
 * virtual addresses, and TIP-20 token addresses.
 *
 * @example
 * ```ts twoslash
 * import { VirtualMaster } from 'ox/tempo'
 *
 * const valid = VirtualMaster.validateSalt({
 *   address: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266',
 *   salt: '0x00000000000000000000000000000000000000000000000000000000abf52baf',
 * })
 *
 * valid
 * // @log: true
 * ```
 *
 * @param value - Master address and salt.
 * @returns `true` if the first 4 bytes of the registration hash are zero.
 */
export function validateSalt(value) {
    try {
        return hasProofOfWork(Hash.keccak256(Hex.concat(resolveAddress(value.address), toFixedHex(value.salt, 32)), { as: 'Bytes' }));
    }
    catch {
        return false;
    }
}
/**
 * Searches a bounded range of salts for the first value that satisfies TIP-1022 PoW.
 *
 * [TIP-1022](https://docs.tempo.xyz/protocol/tips/tip-1022)
 *
 * This is intentionally a small, deterministic primitive. It does not coordinate
 * workers or async execution. Callers that need large searches can shard ranges
 * externally.
 *
 * Master addresses must satisfy TIP-1022 registration constraints: they cannot
 * be the zero address, another virtual address, or a TIP-20 token address.
 *
 * @example
 * ```ts twoslash
 * import { VirtualMaster } from 'ox/tempo'
 *
 * const result = VirtualMaster.mineSalt({
 *   address: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266',
 *   start: 0xabf52ba0n,
 *   count: 16,
 * })
 *
 * result?.salt
 * // @log: '0x00000000000000000000000000000000000000000000000000000000abf52baf'
 * ```
 *
 * @param value - Search range parameters.
 * @returns The first matching salt in the range, if any.
 */
export function mineSalt(value) {
    assertCount(value.count);
    const address = resolveAddress(value.address);
    const addressBytes = Bytes.fromHex(address);
    const saltBytes = toFixedBytes(value.start ?? 0n, 32);
    const input = new Uint8Array(addressBytes.length + saltBytes.length);
    input.set(addressBytes);
    for (let i = 0; i < value.count; i++) {
        input.set(saltBytes, addressBytes.length);
        const registrationHash = Hash.keccak256(input, { as: 'Bytes' });
        if (hasProofOfWork(registrationHash)) {
            return {
                masterId: Hex.fromBytes(registrationHash.subarray(4, 8)),
                registrationHash: Hex.fromBytes(registrationHash),
                salt: Hex.fromBytes(saltBytes),
            };
        }
        if (i < value.count - 1 && !increment(saltBytes))
            break;
    }
    return undefined;
}
function hasProofOfWork(hash) {
    return hash[0] === 0 && hash[1] === 0 && hash[2] === 0 && hash[3] === 0;
}
function assertCount(count) {
    if (Number.isSafeInteger(count) && count > 0)
        return;
    throw new Errors.BaseError(`Count "${count}" is invalid. Expected a positive safe integer.`);
}
function increment(bytes) {
    for (let i = bytes.length - 1; i >= 0; i--) {
        const value = bytes[i];
        if (value === 0xff) {
            bytes[i] = 0;
            continue;
        }
        bytes[i] = value + 1;
        return true;
    }
    return false;
}
function resolveAddress(address) {
    const resolved = TempoAddress.resolve(address);
    Address.assert(resolved, { strict: false });
    assertValidMasterAddress(resolved);
    return resolved;
}
function assertValidMasterAddress(address) {
    const normalized = address.toLowerCase();
    if (normalized === zeroAddress)
        throw new Errors.BaseError('Virtual master address cannot be the zero address.');
    if (VirtualAddress.isVirtual(address))
        throw new Errors.BaseError('Virtual master address cannot itself be a virtual address.');
    if (normalized.startsWith(tip20Prefix))
        throw new Errors.BaseError('Virtual master address cannot be a TIP-20 token address.');
}
function toFixedBytes(value, size) {
    return Bytes.fromHex(toFixedHex(value, size));
}
function toFixedHex(value, size) {
    if (typeof value === 'number' || typeof value === 'bigint')
        return Hex.fromNumber(value, { size });
    if (typeof value === 'string') {
        Hex.assert(value, { strict: true });
        return Hex.padLeft(value, size);
    }
    return Hex.fromBytes(Bytes.padLeft(value, size));
}
