// ox/_cjs/tempo/VirtualMaster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRegistrationHash = getRegistrationHash;
exports.getMasterId = getMasterId;
exports.validateSalt = validateSalt;
exports.mineSalt = mineSalt;
const Address = require("../core/Address.js");
const Bytes = require("../core/Bytes.js");
const Errors = require("../core/Errors.js");
const Hash = require("../core/Hash.js");
const Hex = require("../core/Hex.js");
const TempoAddress = require("./TempoAddress.js");
const VirtualAddress = require("./VirtualAddress.js");
const tip20Prefix = '0x20c000000000000000000000';
const zeroAddress = '0x0000000000000000000000000000000000000000';
function getRegistrationHash(value) {
    return Hash.keccak256(Hex.concat(resolveAddress(value.address), toFixedHex(value.salt, 32)));
}
function getMasterId(value) {
    return Hex.slice(getRegistrationHash(value), 4, 8);
}
function validateSalt(value) {
    try {
        return hasProofOfWork(Hash.keccak256(Hex.concat(resolveAddress(value.address), toFixedHex(value.salt, 32)), { as: 'Bytes' }));
    }
    catch {
        return false;
    }
}
function mineSalt(value) {
    assertCount(value.count);
    const address = resolveAddress(value.address);
    const addressBytes = Bytes.fromHex(address);
    const saltBytes = toFixedBytes(value.start ?? 0n, 32);
    const input = new Uint8Array(addressBytes.length + saltBytes.length);
    input.set(addressBytes);
    for (let i = 0; i < value.count; i++) {
        input.set(saltBytes, addressBytes.length);
        const registrationHash = Hash.keccak256(input, { as: 'Bytes' });
        if (hasProofOfWork(registrationHash)) {
            return {
                masterId: Hex.fromBytes(registrationHash.subarray(4, 8)),
                registrationHash: Hex.fromBytes(registrationHash),
                salt: Hex.fromBytes(saltBytes),
            };
        }
        if (i < value.count - 1 && !increment(saltBytes))
            break;
    }
    return undefined;
}
function hasProofOfWork(hash) {
    return hash[0] === 0 && hash[1] === 0 && hash[2] === 0 && hash[3] === 0;
}
function assertCount(count) {
    if (Number.isSafeInteger(count) && count > 0)
        return;
    throw new Errors.BaseError(`Count "${count}" is invalid. Expected a positive safe integer.`);
}
function increment(bytes) {
    for (let i = bytes.length - 1; i >= 0; i--) {
        const value = bytes[i];
        if (value === 0xff) {
            bytes[i] = 0;
            continue;
        }
        bytes[i] = value + 1;
        return true;
    }
    return false;
}
function resolveAddress(address) {
    const resolved = TempoAddress.resolve(address);
    Address.assert(resolved, { strict: false });
    assertValidMasterAddress(resolved);
    return resolved;
}
function assertValidMasterAddress(address) {
    const normalized = address.toLowerCase();
    if (normalized === zeroAddress)
        throw new Errors.BaseError('Virtual master address cannot be the zero address.');
    if (VirtualAddress.isVirtual(address))
        throw new Errors.BaseError('Virtual master address cannot itself be a virtual address.');
    if (normalized.startsWith(tip20Prefix))
        throw new Errors.BaseError('Virtual master address cannot be a TIP-20 token address.');
}
function toFixedBytes(value, size) {
    return Bytes.fromHex(toFixedHex(value, size));
}
function toFixedHex(value, size) {
    if (typeof value === 'number' || typeof value === 'bigint')
        return Hex.fromNumber(value, { size });
    if (typeof value === 'string') {
        Hex.assert(value, { strict: true });
        return Hex.padLeft(value, size);
    }
    return Hex.fromBytes(Bytes.padLeft(value, size));
}
